File: Greedy/1202번 - 보석 도둑/1202.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int n, k;
vector<pair<int, int> > gem;
vector<int> bag;

void input() {
    cin >> n >> k;
    for(int i = 0, m, v; i < n; i++) {
        cin >> m >> v;
        gem.push_back({m, v});
    }
    for(int i = 0, tmp; i < k; i++) {
        cin >> tmp;
        bag.push_back(tmp);
    }
}

void solve() {
    sort(gem.begin(), gem.end());
    sort(bag.begin(), bag.end());
    
    long long ans = 0;
    int idx = 0;
    
    priority_queue<int> pos_weight;
    
    for(int i = 0; i < k; i++) {
        while(idx < n && gem[idx].first <= bag[i]) {
            pos_weight.push(gem[idx].second);
            idx++;
        }
        
        if(pos_weight.size()) {
            int w = pos_weight.top(); pos_weight.pop();
            ans += w;
        }
    }
    cout << ans << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();

    return 0;
}

File: Greedy/1202번 - 보석 도둑/1202.py
import sys
import heapq

input = sys.stdin.readline

n, k = map(int, input().split())
gem = []
bag = []
for _ in range(n):
    m, v = map(int, input().split())
    gem.append([m, v])
for _ in range(k):
    bag.append(int(input()))
    
def solve():
    gem.sort()
    bag.sort()
    
    ans = idx = 0
    pq = []
    
    for i in range(k):
        while idx < n and bag[i] >= gem[idx][0]:
            heapq.heappush(pq, (-gem[idx][1], gem[idx][1]))
            idx += 1
        if pq:
            ans += heapq.heappop(pq)[1]
            
    print(ans)

if __name__ == '__main__':
    solve()